### PR TITLE
Migration guide for v8 to v9

### LINKED ISSUE #944: Write migration guide for Neutrino 9
We need to add a new section to:
https://github.com/mozilla-neutrino/neutrino-dev/blob/master/docs/migration-guide.md

### PR DESCRIPTION
**Pull request by eliperelman**
Fixes #944.

Work in progress.

### CODE REVIEW

File: docs/migration-guide.md [hunk 1]
```diff
@@ -2,11 +2,196 @@
...
+will tell webpack to use the Neutrino middleware to load its configuration:
+
+```js
+// webpack.config.js
+const neutrino = require('neutrino');
+
+module.exports = neutrino().webpack();
+```
+
+Or for ESLint:
```
Comment (edmorley) [inline docs/migration-guide.md]: The `.eslintrc.js` is something that many people might have already created - they'll need to update it, given the Neutrino API changes

File: docs/migration-guide.md [hunk 2]
```diff
@@ -2,11 +2,196 @@
...
 
+## Neutrino v8 to v9
+
+Neutrino v9 is our largest release ever, bringing the preset and middleware ecosystem back to the
+native tools and utilities for which they were originally created. The biggest breaking change
+is the necessity to use external tools *alongside* Neutrino now, i.e. webpack, ESLint, Jest, Karma,
+and others' native CLIs will be used in tandem with Neutrino. A list of changes is detailed below for
+migrating:
+
+- **BREAKING CHANGE** The minimum supported Node.js version is v8.3. Node.js v9 is no longer supported
```
Comment (edmorley) [inline docs/migration-guide.md]: `s/8.3/8.10/` (#1035)

Also, I think Node.js has dropped the `v` prefix? (Due to confusion with `v8` etc)

File: docs/migration-guide.md [hunk 3]
```diff
@@ -2,11 +2,196 @@
...
 - [Neutrino v4 to v5](#neutrino-v4-to-v5)
 
+## Neutrino v8 to v9
+
+Neutrino v9 is our largest release ever, bringing the preset and middleware ecosystem back to the
+native tools and utilities for which they were originally created. The biggest breaking change
+is the necessity to use external tools *alongside* Neutrino now, i.e. webpack, ESLint, Jest, Karma,
+and others' native CLIs will be used in tandem with Neutrino. A list of changes is detailed below for
+migrating:
+
```
Comment (edmorley) [inline docs/migration-guide.md]: Perhaps we should add a suggestion to refer to the differences between the output of `neutrino {build, ...} --inspect-new` vs `neutrino --inspect --mode {production, ...}` when debugging the upgrade?

File: docs/migration-guide.md [hunk 4]
```diff
@@ -2,11 +2,196 @@
...
+- **BREAKING CHANGE** The `options.mains` Neutrino option can now also accept an object per entry point for setting
+individual options for html-webpack-plugin [#1029](https://github.com/neutrinojs/neutrino/pull/1029). The previous API
+for setting these properties as strings is still supported. This also introduces a breaking API change as accessing an
+entry point property on `neutrino.options.mains` now returns a normalized object, even when using string options.
+- **BREAKING CHANGE** When configuring webpack with a target of `node` the `@neutrinojs/jest` preset will default its
+`testEnvironment` to also be `node` instead of `jsdom` [#1030](https://github.com/neutrinojs/neutrino/pull/1030).
+- **BREAKING CHANGE** Various dependencies have been updated which may bring their own breaking changes. Please
+check and test your project to ensure proper functionality.
+- ESLint caching is now enabled by default, so it is recommended to specify `.eslintcache` as being ignored from
+your source control commits.
```
Comment (edmorley) [inline docs/migration-guide.md]: Additional entries:
* options.hotEntries option has been removed from `@neutrinojs/web` and its dependants (#902)
* The way `root` should be specified has changed (see #1041 for how to set it via `options` instead)
* `worker-loader` has been removed from `@neutrinojs/web` (#1069)
* The `config.resolve.extensions` order has now changed to be closer to that of webpack (#1080)
* The tests directory is linted now too by eslint-loader (#951).

File: docs/migration-guide.md [hunk 5]
```diff
@@ -2,11 +2,196 @@
...
+
```
Comment (edmorley) [inline docs/migration-guide.md]: Should we add a link to the full list of changes? (eg [https://github.com/neutrinojs/neutrino/compare/v8.3.0...master](https://github.com/neutrinojs/neutrino/compare/v8.3.0...master) but updated to use the v9.0.0 tag when it exists)

File: docs/migration-guide.md [hunk 6]
```diff
@@ -2,11 +2,196 @@
 
 ## Contents
 
+- [Neutrino v8 to v9](#neutrino-v8-to-v9)
 - [Neutrino v7 to v8](#neutrino-v7-to-v8)
 - [Neutrino v6 to v7](#neutrino-v6-to-v7)
 - [Neutrino v5 to v6](#neutrino-v5-to-v6)
 - [Neutrino v4 to v5](#neutrino-v4-to-v5)
```
Comment (edmorley) [inline docs/migration-guide.md]: Should we add a link to the issues section saying please file an issue if any problems found when upgrading that aren't mentioned etc / they need help?

File: docs/migration-guide.md (hunk 4 above)
Comment (edmorley) [reply docs/migration-guide.md]: Oh and also:
* That `.neutrinorc.js` is now mandatory (before people could skip it and use `--use`)
* The Neutrino events/lifecycle + several other APIs no longer exist (though this is more for middleware authors, so perhaps we can ignore?)
* `options.command` no longer exists (if people were using a conditional on that, rather than `NODE_ENV`)

File: docs/migration-guide.md [hunk 7]
```diff
@@ -2,11 +2,196 @@
...
+```js
+// webpack.config.js
+const neutrino = require('neutrino');
+
+module.exports = [
+  // first build configuration
+  neutrino().webpack(),
+  
+  // second build configuration
+  neutrino().webpack(config => ({
```
[the remark this answers is not in the document]
Comment (edmorley) [reply docs/migration-guide.md]: I think this means that this comment needs updating too?
https://github.com/neutrinojs/neutrino/issues/1015#issuecomment-420306768

File: docs/migration-guide.md (hunk 4 above)
Comment (edmorley) [reply docs/migration-guide.md]: Also:
* we're now using our own template instead of `html-webpack-template` (#1049)

File: docs/migration-guide.md (hunk 2 above)
Comment (eliperelman) [reply docs/migration-guide.md]: Weird, never heard this, although I thought the big differentiator should have been the casing.

Comment (edmorley) [reply docs/migration-guide.md]: > Note that, when referring to Node.js release versions, we have dropped the "v" in Node.js 8. Previous versions were commonly referred to as v0.10, v0.12, v4, v6, etc. In order to avoid confusion with V8, the underlying JavaScript engine, we've dropped the "v" and call it Node.js 8.

From https://nodejs.org/en/blog/release/v8.0.0/#long-term-support

(I had to look it up to check I wasn't mis-remembering! :-))

File: docs/migration-guide.md (hunk 5 above)
Comment (eliperelman) [reply docs/migration-guide.md]: Sure.

**Comment by eliperelman** [comment]
OK, I think this is now updated from the review.

File: docs/migration-guide.md (hunk 2 above)
Comment (eliperelman) [reply docs/migration-guide.md]: TIL!

File: docs/migration-guide.md [hunk 8]
```diff
@@ -2,11 +2,196 @@
...
+  ]
+}
+```
+
+- **BREAKING CHANGE** Using a test middleware with no `NODE_ENV` set will set `process.env.NODE_ENV`
+to `test` by default [#972](https://github.com/neutrinojs/neutrino/pull/972).
+- **BREAKING CHANGE** Upgraded to latest major versions of webpack, webpack-dev-server, and related
+packages [#809](https://github.com/neutrinojs/neutrino/pull/809).
+- **BREAKING CHANGE** The `@neutrinojs/chunk` middleware has been removed in favor of webpack's
+improved functionality around `splitChunks` [#809](https://github.com/neutrinojs/neutrino/pull/809).
```
[the remark this answers is not in the document]
Comment (edmorley) [reply docs/migration-guide.md]: Think this change is outstanding?

File: docs/migration-guide.md [hunk 9]
```diff
@@ -2,11 +2,196 @@
...
+- **BREAKING CHANGE** The fast-async Babel plugin is no longer used, so async functions will not be de-sugared
+to Promises by default any more [#790](https://github.com/neutrinojs/neutrino/pull/790). This means the preset option
+of `polyfills.async` has been removed.
+- **BREAKING CHANGE** The output format from `neutrino --inspect` has changed and is equivalent to the
+`neutrino --inspect-new` in Neutrino v8 [#928](https://github.com/neutrinojs/neutrino/pull/928).
+- **BREAKING CHANGE** The length of hashes in built filenames has been shortened to 8 characters
+[#930](https://github.com/neutrinojs/neutrino/pull/930). This will
+cause unchanged files to invalidate caches and if you use regex rules to manage web server caches, those must be
+updated.
+- **BREAKING CHANGE** Module resolution now uses the webpack 4 defaults
```
[the remark this answers is not in the document]
Comment (edmorley) [reply docs/migration-guide.md]: Think this change is outstanding?

[the remark this answers is not in the document]
Comment (edmorley) [reply docs/migration-guide.md]: Whoops wrong comment

File: docs/migration-guide.md [hunk 10]
```diff
@@ -2,11 +2,196 @@
...
+[#926](https://github.com/neutrinojs/neutrino/pull/926). This means that the option `neutrino.options.node_modules`
+has been removed.
+- **BREAKING CHANGE** The font and image loader rules have now been consolidated into a single rule for
+`@neutrinojs/font-loader` and another for `@neutrinojs/image-loader` [#858](https://github.com/neutrinojs/neutrino/pull/858).
+- **BREAKING CHANGE** `@neutrinojs/font-loader` now uses `file-loader` instead of `url-loader`
```
[the remark this answers is not in the document]
Comment (edmorley) [reply docs/migration-guide.md]: Think this change is outstanding?

File: docs/migration-guide.md (hunk 5 above)
Comment (edmorley) [reply docs/migration-guide.md]: This is just `[]` for now; should we change this to use `...master` for now, then update later?

Comment (eliperelman) [reply docs/migration-guide.md]: Yeah, that works for me.

File: docs/migration-guide.md [hunk 11]
```diff
@@ -2,11 +2,262 @@
...
+[#902](https://github.com/neutrinojs/neutrino/pull/902).
+- **BREAKING CHANGE** `@neutrinojs/web` and its dependent middleware no longer include `worker-loader` for automatically
+loading `*.worker.js` files [#1069](https://github.com/neutrinojs/neutrino/pull/1069).
+- **BREAKING CHANGE** The loading order for `config.resolve.extensions` has been rearranged to be closer in parity to
+what webpack has configured by default [#1080](https://github.com/neutrinojs/neutrino/pull/1080).
+- **BREAKING CHANGE** The tests directory is now additionally linted by default with `@neutrinojs/eslint` and its
+dependent middleware [#951](https://github.com/neutrinojs/neutrino/pull/951).
+- **BREAKING CHANGE** Various dependencies have been updated which may bring their own breaking changes. Please
+check and test your project to ensure proper functionality.
+- ESLint caching is now enabled by default, so it is recommended to specify `.eslintcache` as being ignored from
```
Comment (edmorley) [inline docs/migration-guide.md]: `s/by default/by default for new projects/` ?

File: docs/migration-guide.md [hunk 12]
```diff
@@ -45,7 +296,8 @@ module.exports = Neutrino({ root: __dirname })
   .call('eslintrc');
 ```
 
-- **BREAKING CHANGE** When building, `file-loader` and `url-loader` will generate file names with pattern
+- **BREAKING CHANGE** When building, [`file-loader`](https://webpack.js.org/loaders/file-loader/) and
```
Comment (edmorley) [inline docs/migration-guide.md]: Ah I think this change was meant to be above?

File: docs/migration-guide.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (eliperelman) [reply docs/migration-guide.md]: Weird, I have this change locally but it's not here.

File: docs/migration-guide.md [hunk 13]
```diff
@@ -45,7 +297,7 @@ module.exports = Neutrino({ root: __dirname })
   .call('eslintrc');
 ```
 
-- **BREAKING CHANGE** When building, `file-loader` and `url-loader` will generate file names with pattern
+- **BREAKING CHANGE** When building, file-loader and url-loader will generate file names with pattern
```
Comment (edmorley) [inline docs/migration-guide.md]: Could you revert this change to the v7->v8 migration section? :-)
